File: 2015/19/rust/src/main.rs
use std::{collections::HashSet, fs};

fn parse(path: &str) -> (Vec<(String, String)>, String) {
    let input = fs::read_to_string(path).expect("Error reading input file");
    let parts: Vec<&str> = input.trim().split("\n\n").collect();
    
    let rules_data = parts[0];
    let molecule = parts[1].to_string();

    let rules: Vec<(String, String)> = rules_data
        .lines()
        .map(|line| {
            let parts: Vec<&str> = line.split(" => ").collect();
            (parts[0].to_string(), parts[1].to_string())
        })
        .collect();
    
    (rules, molecule)
}

fn gen_replacements(molecule: &str, rules: &[(String, String)]) -> usize {
    let mut molecules: HashSet<String> = HashSet::new();

    for (from, to) in rules {
        let mut start_index = 0;
        while let Some(pos) = molecule[start_index..].find(from) {
            let pos = start_index + pos;
            let new_molecule = format!(
                "{}{}{}",
                &molecule[..pos],
                to,
                &molecule[pos + from.len()..],
            );
            molecules.insert(new_molecule);
            start_index = pos + 1;
        }
    }

    molecules.len()
}

fn gen_from_e(molecule: &str, rules: &[(String, String)]) -> i32 {
    let mut steps = 0;
    let mut current = molecule.to_string();

    while current != "e" {
        let mut replaced = false;
        
        for (from, to) in rules {
            if let Some(pos) = current.find(to) {
                current = format!(
                    "{}{}{}",
                    &current[..pos],
                    from,
                    &current[pos + to.len()..]
                );
                steps += 1;
                replaced = true;
                break;
            }
        }
        if !replaced {
            return -1;
        }
    }
    steps
}

fn main() {
    let (rules, molecule) = parse("../input.txt");
    let sep = "=".repeat(20);
    let part1 = gen_replacements(&molecule, &rules);
    let part2 = gen_from_e(&molecule, &rules);
    println!("{} Part 1 {}\nTotal number of unique combinations is: {}", sep, sep, part1);
    println!("{} Part 2 {}\nTotal number of steps required is: {}", sep, sep, part2);
}

File: 2015/19/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Rule struct {
	From string
	To   string
}

func parse(path string) ([]Rule, string) {
	file, err := os.Open(path)
	if err != nil {
		panic(fmt.Sprintf("Error reading input file: %v", err))
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	splitIndex := 0
	for i, line := range lines {
		if line == "" {
			splitIndex = i
			break
		}
	}

	rulesData := lines[:splitIndex]
	molecule := lines[splitIndex+1]

	var rules []Rule
	for _, line := range rulesData {
		parts := strings.Split(line, " => ")
		if len(parts) == 2 {
			rules = append(rules, Rule{From: parts[0], To: parts[1]})
		}
	}

	return rules, molecule
}

func genReplacements(molecule string, rules []Rule) int {
	molecules := make(map[string]struct{})

	for _, rule := range rules {
		startIndex := 0
		for {
			pos := strings.Index(molecule[startIndex:], rule.From)
			if pos == -1 {
				break
			}
			pos += startIndex
			newMolecule := molecule[:pos] + rule.To + molecule[pos+len(rule.From):]
			molecules[newMolecule] = struct{}{}
			startIndex = pos + 1
		}
	}

	return len(molecules)
}

func genFromE(molecule string, rules []Rule) int {
	steps := 0
	current := molecule

	for current != "e" {
		replaced := false
		for _, rule := range rules {
			pos := strings.Index(current, rule.To)
			if pos != -1 {
				current = current[:pos] + rule.From + current[pos+len(rule.To):]
				steps++
				replaced = true
				break
			}
		}
		if !replaced {
			return -1
		}
	}
	return steps
}

func main() {
	rules, molecule := parse("../input.txt")
	sep := strings.Repeat("=", 20)
	part1 := genReplacements(molecule, rules)
	part2 := genFromE(molecule, rules)

	fmt.Printf("%s Part 1 %s\nTotal number of unique combinations is: %d\n", sep, sep, part1)
	fmt.Printf("%s Part 2 %s\nTotal number of steps required is: %d\n", sep, sep, part2)
}
